lib/Attean/IDPQueryPlanner.pm

```diff
@@ -729,6 +727,55 @@ sub-plan participating in the join.
 		return $sum;
 	}
 
+	# The following method returns a factor used to penalize certain types of joins.
+	# It penalizes cartesian joins heavily, but also uses HSP Heuristic 2 (see REFERENCES)
+	sub _penalize_joins {
+		my ($self, $plan) = @_;
+		my $jv			= $plan->join_variables;
+		my @children	= @{ $plan->children };
+		my $mult		   = 1;
+		if (scalar(@$jv)) {
+			if ( all { $_->isa('Attean::Plan::Quad') } @children[0..1]) {
+				my $var = ${$jv}[0]; # We will join on this
+				my @lnodes = $children[0]->values;
+				my @rnodes = $children[1]->values;
+				# Now, find where the join variables are in the triple patterns
+				my %joinpos;
+				for (my $i = 0; $i <= 2; $i++) {
+					if ($lnodes[$i]->does('Attean::API::Variable') && $lnodes[$i]->value eq $var) {
+						$joinpos{l} = $i;
+					}
+					if ($rnodes[$i]->does('Attean::API::Variable') && $rnodes[$i]->value eq $var) {
+						$joinpos{r} = $i;
+					}
+					last if scalar keys(%joinpos) >= 2; # Perhaps a bit premature optimization
+				}
+				my $joinpos = join("", sort values(%joinpos)); # We can now match on this string
+				if ($joinpos eq '12') { # The penalization numbers come mostly out from thin air
```

Comment:
Yeah, good point, now fixed (with a different notation, I think qw makes it hard to read)